openhands-sdk/openhands/sdk/agent/utils.py

```diff
@@ -96,3 +105,101 @@ def fix_malformed_tool_arguments(
                 pass
 
     return fixed_arguments
+
+
+@overload
+def prepare_llm_messages(
+    state: "ConversationState",
+    condenser: None = None,
+    additional_messages: list[Message] | None = None,
+) -> list[Message]: ...
+
+
+@overload
+def prepare_llm_messages(
+    state: "ConversationState",
+    condenser: CondenserBase,
+    additional_messages: list[Message] | None = None,
+) -> list[Message] | Condensation: ...
+
+
+def prepare_llm_messages(
+    state: "ConversationState",
+    condenser: CondenserBase | None = None,
+    additional_messages: list[Message] | None = None,
+) -> list[Message] | Condensation:
+    """Prepare LLM messages from conversation context.
+
+    This utility function extracts the common logic for preparing conversation
+    context that is shared between agent.step() and ask_agent() methods.
+    It handles condensation internally and calls the callback when needed.
+
+    Args:
+        state: The conversation state containing events
+        condenser: Optional condenser for handling context window limits
+        additional_messages: Optional additional messages to append
+        on_event: Optional callback for handling condensation events
+
+    Returns:
+        List of messages ready for LLM completion
+
+    Raises:
+        RuntimeError: If condensation is needed but no callback is provided
+    """
+
+    view = View.from_events(state.events)
+    llm_convertible_events = view.events
+
+    # If a condenser is registered, we need to give it an
+    # opportunity to transform the events. This will either
+    # produce a list of events, exactly as expected, or a
+    # new condensation that needs to be processed
+    if condenser is not None:
+        condensation_result = condenser.condense(view)
+
+        match condensation_result:
+            case View():
+                llm_convertible_events = condensation_result.events
+
+            case Condensation():
+                return condensation_result
+
+    # Convert events to messages
+    messages = LLMConvertibleEvent.events_to_messages(llm_convertible_events)
```

Comment:
We should filter for LLMConvertible events before passing it to events_to_messages
                e for e in state.events if isinstance(e, LLMConvertibleEvent)